File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/trabalhos_praticos/TP_01/questao_17/PalindromoRec.c
#include <stdio.h>
#include <string.h>

int IsFim(char line[]){
    int result = 1;

    if(line[0] == 'F' && line[1] == 'I' && line[2] == 'M'){
        result = 0;
    }

    return result;
}

int Contar(char str[]){
    int count = 0;
    int i = 0;
    char c = str[i];

    while(c != '\0'){
        count++;
        i++;
        c = str[i];
    }

    return count;
}

int VerificarRec(char str[], int start, int end){
    int result = 1;

    if(start < end){
        if(str[start] != str[end]){
            result = 0;
        } else{
            result = VerificarRec(str, start + 1, end - 1);
        }
    }

    return result;
}

int Verificar(char str[]){
    int end = Contar(str) - 1;

    return VerificarRec(str, 0, end);
}

int main(void){
    char line[500];

    scanf("%[^\n]", line);
    while(IsFim(line) != 0){
        if(Verificar(line) != 0){
            printf("SIM\n");
        } else{
            printf("NAO\n");
        }
        scanf(" %[^\n]", line);
    }

    return 0;
}

File: Algoritimo-Estrutruras-de-Dados/Aeds-ll/2025_01/trabalhos_praticos/TP_01/questao_16/PalindromoRec.java
import java.util.*;

public class PalindromoRec{
    public static int IsFim(String str){
        int result = 1;

        if(str.charAt(0) == 'F' && str.charAt(1) == 'I' && str.charAt(2) == 'M'){
            result = 0;
        }

        return result;
    }

    public static int VerificarRec(String str, int start, int end){
        int result = 1;

        if(start < end){
            if(str.charAt(start) != str.charAt(end)){
                result = 0;
                start = end; 
            } else{
                result = VerificarRec(str, start + 1, end - 1);
            }
        }

        return result;
    }

    public static int Verificar(String str){
        return VerificarRec(str, 0, str.length() - 1);
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        String line;

        line = sc.nextLine();
        while(IsFim(line) != 0){
            if(Verificar(line) != 0){
                System.out.println("SIM");
            } else{
                System.out.println("NAO");
            }
            line = sc.nextLine();
        }
        sc.close();
    }
}
